sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala

```diff
@@ -597,12 +599,26 @@ class ParquetFilters(
         createFilterHelper(pred, canPartialPushDownConjuncts = false)
           .map(FilterApi.not)
 
-      case sources.In(name, values) if canMakeFilterOn(name, values.head)
-        && values.distinct.length <= pushDownInFilterThreshold =>
-        values.distinct.flatMap { v =>
-          makeEq.lift(nameToParquetField(name).fieldType)
-            .map(_(nameToParquetField(name).fieldNames, v))
-        }.reduceLeftOption(FilterApi.or)
+      case sources.In(name, values) if pushDownInFilterThreshold > 0 &&
+        values.nonEmpty && canMakeFilterOn(name, values.head) =>
+        if (values.length <= pushDownInFilterThreshold) {
+          values.flatMap { v =>
+            makeEq.lift(nameToParquetField(name).fieldType)
+              .map(_(nameToParquetField(name).fieldNames, v))
+          }.reduceLeftOption(FilterApi.or)
+        } else {
+          sparkSchema.find { f =>
+            if (caseSensitive) f.name.equals(name) else f.name.equalsIgnoreCase(name)
+          }.map(_.dataType) match {
+            case Some(dataType) =>
+              val sortedValues = values.sorted(TypeUtils.getInterpretedOrdering(dataType))
+              createFilterHelper(
+                sources.And(sources.GreaterThanOrEqual(name, sortedValues.head),
+                  sources.LessThanOrEqual(name, sortedValues.last)),
+                canPartialPushDownConjuncts)
```

Comment:
The logic is same to HiveShim.scala#L746-L750.
https://github.com/apache/spark/blob/09bb9bedcd27e08b86d63a6aed90d42ca4c606be/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala#L746-L750

@cloud-fan @dongjoon-hyun @HyukjinKwon  It can be improved by 6.6X in `InSet -> InFilters (values count: 100, distribution: 10)`: 
```
Parquet Vectorized (Pushdown)                      9520           9560          27          1.7         605.3       1.0X
Parquet Vectorized (Pushdown)                      873             885           11        18.0          55.5       6.6X
```

Reply:
ah, then can we turn it into a util method and use it in all the filter pushdown place?

Reply:
ok, Added a new function to `TypeUtils`.